src/Runtime/Distributions/Automata/Automaton.State.cs

```diff
@@ -73,73 +59,85 @@ public State(int index, IEnumerable<Transition> transitions, Weight endWeight)
             /// </summary>
-            [DataMember]
-            public int Index { get; internal set; } // TODO: setter of this property is needed only for the state removal procedure
+            public int Index { get; }
 
             /// <summary>
             /// Gets or sets the ending weight of the state.
             /// </summary>
-            [DataMember]
-            public Weight EndWeight { get; set; }
-            
+            /// <remarks>
+            /// C# compiler disallows to use property setter if it sees that <see cref="States"/> instance is a temporary.
+            /// It is not smart enough to understand that property setter actually changes something behind a reference.
+            /// To overcome this issue special <see cref="SetEndWeight"/> method is added calling which is equivalent
+            /// to calling property setter but is not rejected by compiler.
+            /// </remarks>
+            public Weight EndWeight
+            {
+                get => this.Data.EndWeight;
+                set => this.Data.EndWeight = value;
+            }
+
             /// <summary>
-            /// Gets a value indicating whether the ending weight of this state is greater than zero.
+            /// Sets the <see cref="EndWeight"/> property of State. See <see cref="EndWeight"/> remarks on to
+            /// why property setter can not always be used.
             /// </summary>
-            public bool CanEnd
+            /// <param name="weight">New end weight.</param>
+            public void SetEndWeight(Weight weight)
             {
-                get { return !this.EndWeight.IsZero; }
+                this.EndWeight = weight;
             }
 
+            /// <summary>
+            /// Gets a value indicating whether the ending weight of this state is greater than zero.
+            /// </summary>
+            public bool CanEnd => this.Data.CanEnd;
+
             /// <summary>
             /// Gets the number of outgoing transitions.
             /// </summary>
-            public int TransitionCount
-            {
-                get { return this.transitionCount; }
-            }
+            public int TransitionCount => this.Data.TransitionCount;
 
             /// <summary>
             /// Creates the copy of the array of outgoing transitions. Used by quoting.
             /// </summary>
             /// <returns>The copy of the array of outgoing transitions.</returns>
-            public Transition[] GetTransitions()
-            {
-                var result = new Transition[this.transitionCount];
-                Array.Copy(this.transitions, result, this.transitionCount);
-                return result;
-            }
+            public Transition[] GetTransitions() => this.Data.GetTransitions();
+
+            /// <summary>
+            /// Compares 2 states for equality.
+            /// </summary>
+            public static bool operator ==(State a, State b) => a.Data == b.Data;
```

Comment:
Is this correct? Don't owner and index matter here?

Reply:
Because state can belong only to one automaton. (State objects are not shared between atomatons) - we don't have to compare Owner or Index. But I will change this code in future.